### PR TITLE
Post MultiCapture Merge Cleanup/Fixes

### PR DESCRIPTION
**Pull request by gabizou**
Some small bugs have cropped up since merging #2163 .

This PR rewrites a majority of the `BlockTransaction` system devised for #2163. The issues brought up by #2163 involve several neighbor notifications being muted for intermediary block changes (since `BlockSnapshot`s were processed only as an "original" and "final" comparison, eliminating any of the `BlockChangeFlag`s that could have been associated between the original and final state. Likewise, because `TileEntity` instances were being proxied, and the `IBlockState` instances were proxied for specific positions, contamination from the final `BlockTransaction` being captured versus the first `BlockTransaction` being played back and simulated for it's changes could end up causing invalid `IBlockState`s being proxied at specific positions, such that `TileEntity` instances could remove themselves prematurely, or incorrect `IBlockState`s could be called to notify neighbors.

The design is aimed for high speed and efficiency, minimal lookups to the world, using the proxy to fetch/check against for `TileEntity` instances, `IBlockState`s, and validation between transactions as far as "what existed before" for various positions to avoid potential contamination during specific `BlockTransaction` processing (Playing back a neighbor notification after a change block that performed other neighbor notifications, affecting the next block over).

So far, the bugs aimed to be resolved are:
- [x] https://github.com/SpongePowered/SpongeForge/issues/2639
- [x] https://github.com/SpongePowered/SpongeCommon/issues/2212
- [x] https://github.com/SpongePowered/SpongeCommon/issues/2218
- [x] https://github.com/SpongePowered/SpongeForge/issues/2641
- [x] https://github.com/SpongePowered/SpongeForge/issues/2644
- [x] https://github.com/SpongePowered/SpongeForge/issues/2646
- [x]  https://github.com/SpongePowered/SpongeForge/issues/2653
- [x] https://github.com/SpongePowered/SpongeForge/issues/2654

There's probably more issues that can be collected with this PR, but the ones listed above are the first priority to be resolved by this PR.

### CODE REVIEW

**Comment by XakepSDK** [comment]
- [ ] https://github.com/SpongePowered/SpongeForge/issues/2653

File: src/main/java/org/spongepowered/common/event/tracking/PhaseTracker.java [hunk 1]
```diff
@@ -103,8 +103,80 @@
...
+    public static final String DEDICATED_SERVER = "net.minecraft.server.dedicated.DedicatedServer";
+    public static final String MINECRAFT_SERVER = "net.minecraft.server.MinecraftServer";
+    public static final String INTEGRATED_SERVER = "net.minecraft.server.integrated.IntegratedServer";
+    @Nullable private Thread sidedThread;
+    private boolean hasRun = false;
+
+
+    @SuppressWarnings("ThrowableNotThrown")
+    public void setThread(@Nullable Thread thread) throws IllegalAccessException {
+        final StackTraceElement[] stackTrace = new Throwable().getStackTrace();
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/common/event/tracking/PhaseTracker.java:152]: `Thread.currentThread().getStackTrace();`?

**Comment by gabizou** [comment]
For those brave enough to test, please use this jar: https://www.mediafire.com/file/2onnpld06indg0z/spongeforge-1.12.2-2768-7.1.6-RC0.jar/file to test against this PR.

**Comment by gabizou** [comment]
Tested and re-tested with builds and added improvements, could not reproduce the issues outlined above.

**Comment by TheSpicePhantom** [comment]
Version before: spongeforge-1.12.2-2768-7.1.6-RC3629.jar
Version after / FIX: spongeforge-1.12.2-2768-7.1.6-RC3616.jar

Mods that didn't work correctly: 
 - CC: Tweaked
 - Project Red
 - Maybe some more, didn't test all.

What didn't work:

- CC: Tweaked
  - Turtles and Computers would reset after 2 seconds of running their programs

- Project Red
  - Timer:
    - It works, but doesn't output signal to any adjacent cables or redstone dust



FIXED with RC3616


Wanted to point to doot, on discord who helped me figure it out.

**Pull request merged**